### ISSUE TITLE
Test our DSL for creating real world bill payment forms

### ISSUE THREAD

**Issue by djpnewton** [zap-me/zap_bill_payment#6]
Lets gather the requirements for bank payments to some NZ utilities and check that our current system is flexible enough to support them (we can also reuse these for demonstration/prototyping)

- Create a utility database with the mobile phone utilities (vodafone, spark, 2degrees, skinny) and some power companies (genesis, contact, mercury, meridian)
- share the data (maybe we put it into this issue,  wiki page or maybe add it into the code so it seeds the database at startup)
- then we can discuss if there are any issues with our  DSL (https://github.com/zap-me/zap_bill_payment/blob/master/utility_form_dsl.md) or if we need to update it to be more flexible

**Comment by eoliveros** [comment]
Utility Bank Information(Will keep editing while looking for info):
* Mobile:
`-` SPARK:
```
Account Number.: 01-1820-0000123-000
Reference: Spark account number
Code: {Anything customer wants?}

Overseas customers:
Swift code: ANZBNZ22
Bank Address: ANZ Centre, 23-29 Albert Street, Auckland.
```
`-` VODAFONE(has 4 different bank account, depending on the bill they'll be asked to choose one to make payments):
```
Vodafone:
Bank account name	Vodafone A
Account	02-0248-0228009-024
Particulars	Your name
Code	Your Vodafone account number
Reference	Your phone number
SWIFT (if required)	BKNZNZ22


Bank account name	Vodafone B
Account	01-0102-0108338-001
Particulars	Your name
Code	Your Vodafone account number
Reference	Your phone number
SWIFT (if required)	ANZBNZ22


Bank account name	Vodafone C
Account	02-0100-0074189-000
Particulars	Your name
Code	Your Vodafone account number
Reference	Your phone number
SWIFT (if required)	BKNZNZ22

Bank account name	Vodafone D
Account	02-0108-0261413-00
Particulars	Your name
Code	Your Vodafone account number
Reference	Your phone number
SWIFT (if required)	BKNZNZ22
```

`-` 2DEGREES(Found the details on "Direct Debit Authority Terms and conditions" in broadband page. Maybe different for mobile) :
```
Account Number.: 02-0820-0188002-000
Reference Number: Phone number 
Code: Account number
```
Our bank account number with BNZ is 02-0820-0188002-000, **please use this number with your account number and phone number as a reference** if you intend to make online payments for debts incurred to Two Degrees New Zealand Limited


`-` SKINNY(can't find the info for skinny):
```
```
We like to keep things simple, which is why Skinny Direct only works with a registered credit card or debit card issued by MasterCard or Visa. This means you can’t use  Skinny Mobile top up vouchers or any other methods of payment.

**Comment by djpnewton** [comment]
ok we need this formatted for our project

**Comment by eoliveros** [comment]
Trying to understand, should i be substituting values in the DSL and see if it works? not entirely sure, what i'm supposed to do now.

**Comment by djpnewton** [comment]
1. create the utility entries in the heroku instance i shared with you http://zap-bill-payment.herokuapp.com/admin/
2. test the results http://zap-bill-payment.herokuapp.com/utilities

**Comment by eoliveros** [comment]
created the entries in uitility. Then tried creating a payment, the result is the image:
![image](https://user-images.githubusercontent.com/47799652/87387947-de932280-c5f7-11ea-94ad-df8e40e35c60.png)


Its failing at the moment. Something about authentication.

**Comment by djpnewton** [comment]
fixed the bronze authentication

1. why have you not updated the 'fields_description' for the utilities?
2. do the power utilities also (as asked initially)

**Comment by eoliveros** [comment]
> * why have you not updated the 'fields_description' for the utilities?

need clarification on step 1. Please look at heroku to see if that's what you meant for update field_description.

updating step 2 now.

**Comment by djpnewton** [comment]
yes, for step 1 I meant we want real data in the fields_description that actually matches how the utility expects to be paid

can you also add a list of links for where you got the data from so we can double check if need be

**Comment by eoliveros** [comment]
> yes, for step 1 I meant we want real data in the fields_description that actually matches how the utility expects to be paid
> 
> can you also add a list of links for where you got the data from so we can double check if need be

i'll list them here instead of the app:
```
- spark: https://www.spark.co.nz/help/account/bill/pay-bill/
- vodafone: https://help.vodafone.co.nz/app/answers/detail/a_id/17702
- 2degrees: Can't find this for mobile. Found the broadband side of 2degre3es and thats where the info im using:
   https://www.2degreesmobile.co.nz/termsofuse/broadband/broadband/direct-debit-authority-terms-and-conditions/#:~:text=Our%20bank%20account%20number%20with,Two%20Degrees%20New%20Zealand%20Limited.
- skinny: can't find any info on this

- contact energy: https://contact.co.nz/residential/billing-and-payments#Pay-via-internet-banking
- genesis energy: https://www.genesisenergy.co.nz/billing-and-payments
- mercury energy: https://www.mercury.co.nz/help/faq/how-to-pay-my-bill
- meridian energy: https://www.meridianenergy.co.nz/business/account-and-support/pay-bill/internet-banking
```

**Comment by djpnewton** [comment]
I have updated the DSL (see https://github.com/zap-me/zap_bill_payment/issues/7)

See the updated document (https://github.com/zap-me/zap_bill_payment/blob/master/utility_form_dsl.md) and update the `bank_description` fields to match. You should be able to now have the 4 vodafone bank accounts in the same utility and also include a link in the description to the vodafone page which tells the customer how to figure out which bank account to use

**Comment by eoliveros** [comment]
> I have updated the DSL (see #7)
> 
> See the updated document (https://github.com/zap-me/zap_bill_payment/blob/master/utility_form_dsl.md) and update the `bank_description` fields to match. You should be able to now have the 4 vodafone bank accounts in the same utility and also include a link in the description to the vodafone page which tells the customer how to figure out which bank account to use

done the following for the description:
```
[
  {
    "name": "Default Bank (only shown if there are multiple banks)",
    "account_number": "02-0248-0228009-024",
    "fields": [
      {
      "label": "Invoice Number",
      "description": "Vodafone A invoice number on your bill statement",
      "type": "number",
      "min": 10000,
      "max": 99999,
      "allow_empty": false,
      "target": "particulars"
      },
      {
      "label": "Vodafone Account Number",
      "description": "Users Vodafone account number",
      "type": "text",
      "min_chars": 4,
      "allow_empty": false,
      "target": ["code"]
      },
      {
      "label": "Vodafone Phone Number",
      "description": "Users Vodafone account number",
      "type": "text",
      "min_chars": 4,
      "allow_empty": false,
      "target": ["reference"]
      }
    ]
  }
]
```

And basically changed the account_number for Vodafone B, C, and D to match they're corresponding bank account number.

But when I try to visit the page:
```
https://zap-bill-payment.herokuapp.com/utility?utility=2
```
Getting the following:
![image](https://user-images.githubusercontent.com/47799652/87746699-56a25800-c845-11ea-81e6-03cf62b3da7a.png)

What am I missing in here? I can't see a way of changing in the page that would allow the user to swap between A,B,C, and D.

**Comment by djpnewton** [comment]
you need to condense the four vodafone banks into a single utility, I have started the process to give you the idea

http://zap-bill-payment.herokuapp.com/utility?utility=2

**Comment by eoliveros** [comment]
Since this is condensing the no. of utilities, im deleting some of them after they've been condensed to other pages.
